Improve/fix the regular expressions in endf.pyx

furutaka opened the pull request:
The CONTENTS_R regular expression in endf.pyx is not so restrictive, and for example the last line of 
the following, excerpted from n-058_Ce_136.endf file of ENDF/B-VII.1, is thought of as a contents line, causing an error:

>       Collective levels used in direct calcualtions                   5825 1451  138
>                                                                       5825 1451  139
>         Ncoll  Lmax IDef  Kgs  (Def(1,j),j=2,IDef,2)                  5825 1451  140
>           12    4    4   0.0  0.170E+00  0.000E+00                    5825 1451  141

Other two regular expressions were also improved (IMHO) in the same way.

furutaka committed:
Merge pull request #1 from pyne/develop

Bringing fork up to date

Commit:
Improve regular expressions in endf.pyx

The original regular expression to pick up the contents lines in (MF,MT)=(1,451),
  CONTENTS_R = re.compile(' +\d{1,2} +\d{1,3} +\d{1,10} +')
is not so restrictive enough that the 141st line in n-058_Ce_136.endf of
ENDF/B-VII.1,

  Collective levels used in direct calcualtions                   5825 1451  138
                                                                  5825 1451  139
    Ncoll  Lmax IDef  Kgs  (Def(1,j),j=2,IDef,2)                  5825 1451  140
      12    4    4   0.0  0.170E+00  0.000E+00                    5825 1451  141

is recognized as a contents line and causes an error:

   >>> ds=ENDFDataSource("n-058_Ce_136.endf")
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
     File "/home/furutaka/.local/lib/python3.7/site-packages/pyne/xs/data_source.py", line 739, in __init__
       self.library = endf.Library(fh)
     File "endf.pyx", line 80, in pyne.endf.Library.__init__
     File "endf.pyx", line 183, in pyne.endf.Library._read_headers
   ValueError: invalid literal for int() with base 10: '.0  0.170E+'

According to the "ENDF-6 Formats Manual"
  https://www.nndc.bnl.gov/csewg/docs/endf-manual.pdf#page=58
a contents line is one of CONT Records and should consist of the leading 22
whitespaces and following four (at most 11-digit) integers.

Taking the above specification into account, the new regular expression,
  CONTENTS_R = re.compile('^ {22}([ \d]{10}\d){4}')
is more restrictive and do not cause the same kind of errors as above, although
it is NOT PERFECT because it can match the following line (the first two lines are
just for the ease of counting):
(         1          2          3          4          5          6              )
(2345678901234567890123456789012345678901234567890123456789012345678901234567890)
                      1 2 3 4 5 67 8 9 0 1 23 4 5 6 7 89 0 1 2 3 45825 1451  141

The same error occurs with at least the following three ENDF/B-VII.1
files: n-077_Ir_191.endf, n-077_Ir_193.endf, n-081_Tl_205.endf

In the same way, two other regular expressions, FILE1_R and NUMERICAL_DATA_R
were modified.

One of the duplicated SPACE66_R was deleted.

gonuke commented:
Thanks for this contribution @furutaka!  It is always a pleasure to receive new input from users.

I have two requests for updates to this modification:
1. Can you please add a `news` file?  You can see examples in the `news` directory from previous pull requests.  This helps us ensure that we can explain all the changes as we publish updates.

2. Can you add one or more tests?  Such tests will ideally demonstrate the failure using the old code and then demonstrate success using your update.  You may need to add a file to `tests/file_test_endf` as well as adding/modifying methods in `tests/test_endf.py`

Commit:
Added a news file and a test for the CONTENTS_R regexp fix.

gonuke commented on pyne/endf.pyx:
```diff
@@ -40,11 +40,10 @@ libraries = {0: 'ENDF/B', 1: 'ENDF/A', 2: 'JEFF', 3: 'EFF',
-CONTENTS_R = re.compile(' +\d{1,2} +\d{1,3} +\d{1,10} +')
-SPACE66_R = re.compile(' {66}')
-NUMERICAL_DATA_R = re.compile('[\d\-+. ]{80}\n$')
+FILE1_R = re.compile(r'1451[ \d]{4}\d$')
```
This still isn't quite right.

The old one will match incorrectly on  `1451  5` - not enough spaces and/or digits.
This one will match incorrectly on `1451  5 5` - a mixture of the correct number of spaces or digits.
Perhaps it should be:
```suggestion
FILE1_R = re.compile(r'1451(?=[ ]*\d+$)[ \d]{5}$')
```
This adds a look head to ensure that there is one group of 0+ spaces and the one group of 1+ digits before the new line, and then makes sure there are 5 of them, combined.

gonuke replied:
It might be useful to add tests for these regex expressions that don't rely on a file, but simply document a set of developer-constructed cases that either should or should not match and then a test function that shows if/when they do/don't.

furutaka replied:
Let me sleep on it...

gonuke reviewed (changes requested):
I think your expressions are definitely an improvement, but since we are trying to use regex to match FORTRAN fixed format numbers, it's still not right.  I learned something useful here: https://stackoverflow.com/questions/5116422/regex-to-read-fixed-width-numeric-fields

Maybe your new expressions can be updated along these lines?

furutaka committed:
Improve further regular expressions in endf.pyx.

The way (MF,MT)=(1,451) lines are searched for was modified;
by matching a line as a whole (from the beginning to the end
of line).

On the other hand, by scanning a line as a whole, it is NOT
easy to correctly tell what the contents of a line are.
Apparently, the part of a line left to (MAT,MF=1,MT=451) can
be divided into six parts each with 11 characters if the line
is Data or Contents one (CONT Record of ENDF-102).
Therefore, the six strings are tried to match with
  two "E-less" floating numbers and four I11 intergers (Data)
or
  22 spaces followed by four I11 intergers (Contents).
All lines not matched with are treated as those for Document.

Commit:
test_contents_regexp() in tests/test_endf.py modified.

Now it does not rely on any external files; tested with
lines of a string.

The news file was also updated.

Commit:
Allow negative ints but prohibit digits w/ leftmost zero except for single "0"

The LRP flag can be -1.

gonuke commented:
This is definitely looking a little cleaner - I like the idea of breaking each line into blocks of 11 characters.

However, I suddenly remembered that we already have some generic FORTRAN fixed-format reading capability: https://github.com/pyne/pyne/blob/develop/pyne/fortranformat/FortranRecordReader.py
I wonder if it's worth retooling for that?

furutaka commented:
Well... It seems to me that the FortranRecordReader.py is a tool for reading/converting values after knowing what a line is for.  But the problem I had was incorrect recognition of the Contents lines in MT=451, i.e. how to know/tell the Contents lines from others.

furutaka committed:
Further simplify/correct endf.Library._read_headers()

It is no need to detect DATA lines because these exist only
at the beginning of the (MF,MT)=(1,451) lines and these are
already read outside the `while FILE1_R.search(line)` loop.
Therefore the lines for its detection as well as regexp
endf.ELESSFLOAT_R and the method endf.Library._isDataLine()
were removed.

The endf.Library.line_length includes those of newline
character(s) ('\n', '\r', or '\r\n'), but with the universal
newline mode of Python (enabled by default in recent Python)
these are all translated into '\n'.  Therefore, the reading
for matflagstring, fh.read(3*self.line_length), may go TOO
FURTHER; the last three lines are now read by
fh.read(3*len(line)).

After reading the four DATA lines at the beginning, there is
no need to "go back to line after first line."

the length of a line in all ENDF-6 compliant
data files is always 81 because

furutaka committed:
Updated news/fix_endf_regexp.rst.

furutaka committed:
Updated tests/test_endf.py.

gonuke commented on pyne/endf.pyx:
```diff
@@ -152,18 +162,14 @@ class Library(rxdata.RxLib):
-        pos = fh.tell()
         # check for isomer (LIS0/LISO entry)
-        matflagstring = line + fh.read(3*self.line_length)
+        matflagstring = line + fh.read(3*len(line))
```
Is there a reason to not use the member variable here?  The purpose of this variable is to be robust across Window vs Mac vs Linux.

furutaka replied:
Well... I'm trying to use pyne.endf to read JENDL4 files.  Unfortunately, some of which are with '\r\n' and others have '\n', and the reading failed for files w/ '\r\n'.  So I followed the function of Library._read_headers() and found that for the files w/ '\r\n' the fh.read(3*self.line_length) goes too further!  I think that it is because of the universal newline mode; by experimenting, it seems to me that the size argument of read() is not the number of characters in the file but the characters it returns, although I'm not quite sure...

furutaka replied:
##### Sr087.dat: ASCII text
_read_tpid: chars_til_now =  81
_read_tpid: line_length   =  81
_read_tpid: offset        =  0
_read_tpid: len(line)     =  81
##### Sr088.dat: ASCII text, with CRLF line terminators
_read_tpid: chars_til_now =  82
_read_tpid: line_length   =  82
_read_tpid: offset        =  0
_read_tpid: len(line)     =  81


This way, regardless of the newline characters (`'\r\n'` or `'\n'`) and the value of `self.line_length`, `len(line)` always returns 81, and `self.offset` is always 0.

`self.chars_til_now = len(line) + self.line_length - 81`
`self.offset = self.line_length - self.chars_til_now`

furutaka replied:
It seems to me that the same thing (reading too much characters) occurs in endf.Library._read_nucmfmt(), and I'm wondering how to fix them...

gonuke commented on pyne/endf.pyx:
```diff
@@ -175,27 +181,24 @@ class Library(rxdata.RxLib):
-            elif SPACE66_R.match(line):
-                self.structure[nuc]['docs'].append(line[0:66])
-                line = fh.readline()
-            elif NUMERICAL_DATA_R.match(line):
```
Why don't we need this case any more?

furutaka replied:
My understanding is that there're three parts in a (MF,MT)=(1,451) file, the data part which comes first, then the document part, and at last the contents/dictionary part.  As the meaningful values in the data part has already read with fh.read(3*len(line)) (and I don't think we have to go back to the top to read the same data lines again), we only have to distinguish the contents lines from others, i.e. 'docs' lines.

furutaka replied:
Ah, according to the [manual](https://www.nndc.bnl.gov/csewg/docs/endf-manual.pdf#page=54), there're only two parts, 1. a documentation, and 2. a dictionary, and 

> The first three records of the descriptive information contain a standardized presentation
of information on the material, projectile, evaluators, and modification status.

gonuke reviewed (commented):
I really like how this is cleaned up by your changes.

I'm not an expert on ENDF file formats (perhaps we can get some help from @paulromano or someone else on this).

I have a few questions in this review to help gain confidence without my expertise.

furutaka committed:
Fix issues on newline character(s) in pyne.endf.Library

The length of a line in the ENDF-6 format files is 81 or
82 (including the end-of-line character(s)), depending on
the setting of the universal newline mode in Python;
The size of read(size) only counts the number of "translated"
newline characters.

On the other hand, the offset of seek(offset) counts the
RAW newline characters (i.e. "untranslated").

To fix the issues, the definition of some of the attributes
have been changed:

- self.chars_til_now :
          offset from the top of the file
          (newline characters : untranslated)

- self.line_length :
          length of a line read e.g. through readline()
          (newline characters : after translated)

- self.offset :
          difference between self.line_length and the length
          of a line in the ENDF-6 format file

- start in self._read_headers() :
          the first line number of a MAT

- self.mat_dict[nuc]['mfs'][mf, mt] :
          (offset byte of the (MF,MT) from the beginning of the
	   ENDF-6 format file (newline characters untranslated),
	   the length of the (MF,MT) in the size of read() (newline
	   characters translated))

The use of the attributes were changed correspondingly.

furutaka commented:
What do you think of this?

gonuke commented:
Unfortunately, this is now failing the tests for the ENDF reader.  

(As I look at your changes, it seems like perhaps the underlying problem was the mixing of `read()` and `readline()` in the original code.  I think the original author may not have had an appreciation of FORTRAN fixed format and that the there might be a more straightforward way to do this whole module!)

furutaka commented:
The cause of the failure is obvious; the first line of the test file, sampletape1, has 75 characters only!  This is the first time I saw an ENDF-format evaluated nuclear data file with <80 characters...

Can I correct the test file?

By the way, I made a mistake and I'll correct it.

furutaka committed:
pyne.endf.Library.offset can be -1

It is in principle possible that someone pass a filehandle
which opened a file with '\r' or '\n' newline characters
with newline='\r\n', to pyne.endf.Library().

gonuke commented:
I submitted a PR for that incorrect test file.

furutaka committed:
Merge pull request #2 from pyne/develop

Bringing for up to date

furutaka committed:
Merge pull request #3 from pyne/develop

Bringing fork up to date

bam241 commented:
thank you @furutaka !

gonuke commented:
I was just writing a message to merge this since it is definitely an improvement!  Thanks @furutaka!

However, I also think there is room for an overhaul of the ENDF reader since I think the overall approach is not well matched to the fixed format FORTRAN expectations.  It is not easy to parse or maintain this code.  I'll add to issue #720 that perhaps first hinted at this need.
